include/transmem.h

```diff
@@ -1,15 +1,212 @@
+     * @param tstamp timestamp at what time the returned transformation is valid
+     * @return matrix encoding the requested transformation
+     * @throws InvalidArgumentException
+     * @throws NoSuchLinkFoundException
+     */
+    QMatrix4x4 getLink(const FrameID &srcFrame, const FrameID &destFrame, const Timestamp &tstamp);
+
+    /**
+     * @fn QQuaternion getLink(const FrameID &srcFrame, const FrameID &fixFrame, const FrameID &destFrame, const Timestamp &tstamp1, const Timestamp &tstamp2) const
+     * @brief Returns the transformation between @a srcFrame at time @a tstamp1 to @a destFrame at time @a tstamp2.@n The query requires a third frame @a fixFrame which is
+     * stationary relative to @a srcFrame and @a destFrame.
+     * @see TransMem::getLink(const FrameID &srcFrame, const FrameID &destFrame, const Timestamp &tstamp) const
+     * @param srcFrame identifier of the source frame
+     * @param fixFrame identifier of the fix frame
+     * @param destFrame identifier of the destination frame
+     * @param tstamp1 timestamp when the transformation from the source to the fix frame is valid
+     * @param tstamp2 timestamp when the transformation from the fix frame to the destination frame is valid
+     * @return matrix encoding the requested transformation
+     * @throws InvalidArgumentException
+     * @throws NoSuchLinkFoundException
+     */
+    QMatrix4x4 getLink(const FrameID &srcFrame, const FrameID &fixFrame, const FrameID &destFrame, const Timestamp &tstamp1, const Timestamp &tstamp2);
+
+    /**
+     * @fn QQuaternion getBestLink(const FrameID &srcFrame, const FrameID &destFrame, Timestamp &tstamp) const
+     * @brief Returns the best transformation betwen @a srcFrame and @a destFrame.@n@n
+     * Best transformation:@n
+     * Let L1,..,Ln be the links required for the transformation from @a srcFrame to @a destFrame.@n
+     * Let Eij be the j-th transformaton entry inserted at time tij in the link Li.@n
+     * The best transformation is the transformation obtained at time t, were t minimizes the sum over all |t-tij| from 1 to n.@n@n
+     * The resolution of t is 10 ms.
+     * @param srcFrame identifier of the source frame
+     * @param destFrame identifier of the destination frame
+     * @param tstamp returns at what time the best transformation was available
+     * @return matrix encoding the requested transformation
+     * @throws InvalidArgumentException
+     * @throws NoSuchLinkFoundException
+     */
+    QMatrix4x4 getBestLink(const FrameID &srcFrame, const FrameID &destFrame, Timestamp &tstamp);
+
+    void dumpAsJSON();
+
+    void dumpAsGraphML();
+
+protected:
+
+    bool shortestPath(Path &p);
+
+    bool calculateBestPointInTime(Path &p, Timestamp &tStampBestPointInTime);
+
+    bool calculateTransformation(const Path &p, StampedTransformation &e);
+
+    std::unordered_map<FrameID, Frame*> frameID2Frame;
+
+    std::deque<Frame> frames;
```

Comment:
Do you use deque to garbage collect old frames? I am not sure frames will get removed in the same order as links. Also, I am not even sure they should be removed at all. Why not using vector? Or maybe we not even need this container because `frameID2Frame` is enough. Maybe the later could contain the frame object directly? (not just the pointer). Also if frames are never deleted one can store a reference in a link, as it cannot be null, further simplifying the code.